### PR TITLE
OrchardZSA backward compatability using traits

### PR DESCRIPTION
**Pull request by YaoJGalteland**
This PR contains the following modifications:
- Add tests for Lookup, ECC, Merkle, and Sinsemilla to verify that the verification key and the proof have not been modified by comparing them to those saved in a file.
- Introduce a `LookupRangeCheck` trait that provides common methods for a lookup range check.
- Use this new trait as a generic parameter in configs, chips and tests that are using lookup.
- Create a new Lookup chip which is optimized for 4, 5 and 10-bit range check and add tests for this new chip.
- Add `init_from_private_point` parameter in `SinsemillaConfig`. If this parameter is set to false, the `SinsemillaChip` is the same as the current `SinsemillaChip` used in vanilla circuit. Otherwise, the `SinsemillaChip` is modified to support hash from private point.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -52,8 +52,10 @@ jobs:
         with:
           beta-features: ${{ matrix.stage == 'beta' }}
           nightly-features: ${{ matrix.stage == 'nightly' }}
+      - name: Update apt
+        run: sudo apt-get update
       - name: Install cross-platform support dependencies
-        run: sudo apt install gcc-multilib
+        run: sudo apt-get install gcc-multilib
```
Comment (PaulLaux) [inline .github/workflows/ci.yml:58]: what are these changes?

File: halo2_gadgets/src/ecc.rs [hunk 2]
```diff
@@ -967,7 +988,38 @@ pub(crate) mod tests {
...
 
-        let circuit = MyCircuit { test_errors: false };
+        let circuit = MyCircuit::<PallasLookupRangeCheckConfig>::new(false);
+        halo2_proofs::dev::CircuitLayout::default()
+            .render(13, &circuit, &root)
+            .unwrap();
+    }
+
+    #[test]
+    fn ecc_chip_4_5_b() {
```
Comment (PaulLaux) [inline halo2_gadgets/src/ecc.rs:998]: the name should probably be `ecc_chip_4_5b()` for consistency (all)

File: halo2_gadgets/src/ecc/chip/mul_fixed/short.rs [hunk 3]
```diff
@@ -462,272 +470,290 @@ pub mod tests {
...
-
-        #[derive(Default)]
-        struct MyCircuit {
-            magnitude: Value<pallas::Base>,
-            sign: Value<pallas::Base>,
-            // For test checking
-            magnitude_error: Value<pallas::Base>,
-        }
+    #[derive(Default)]
+    struct MyMagnitudeSignCircuit<Lookup: PallasLookupRangeCheck> {
```
Comment (PaulLaux) [inline halo2_gadgets/src/ecc/chip/mul_fixed/short.rs:474]: think `My` is redundant after rename

File: halo2_gadgets/src/sinsemilla/merkle.rs [hunk 4]
```diff
@@ -391,9 +433,200 @@ pub mod tests {
...
+        let root = BitMapBackend::new(
+            "merkle-path-with-private-init-4-5-b-layout.png",
+            (1024, 7680),
+        )
+        .into_drawing_area();
+        root.fill(&WHITE).unwrap();
+        let root = root.titled("MerkleCRH Path", ("sans-serif", 60)).unwrap();
+
+        let circuit: MyCircuitWithHashFromPrivatePoint<PallasLookupRangeCheck45BConfig> =
+            MyCircuitWithHashFromPrivatePoint {
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla/merkle.rs:622]: use `new()` here?

File: halo2_gadgets/src/utilities/cond_swap.rs [hunk 5]
```diff
@@ -412,32 +415,37 @@ mod tests {
...
 
-        impl Circuit<pallas::Base> for MyCircuit {
-            type Config = MyConfig;
+        impl<Lookup: PallasLookupRangeCheck> Circuit<pallas::Base> for MyMuxCircuit<Lookup> {
+            type Config = MyMuxConfig<Lookup>;
             type FloorPlanner = SimpleFloorPlanner;
 
             fn without_witnesses(&self) -> Self {
-                Self::default()
+                MyMuxCircuit::<Lookup> {
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/cond_swap.rs:443]: we should use `new()` for consistency

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 6]
```diff
@@ -478,295 +705,380 @@ mod tests {
...
 
+    use crate::{
+        sinsemilla::primitives::K,
+        tests::test_utils::test_against_stored_circuit,
+        utilities::lookup_range_check::{
+            LookupRangeCheck, PallasLookupRangeCheck, PallasLookupRangeCheck45BConfig,
+            PallasLookupRangeCheckConfig,
+        },
+    };
     use std::{convert::TryInto, marker::PhantomData};
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:717]: can we restructure to reduce diff size?

**Review by PaulLaux — COMMENTED** [review]
Good overall, added some minor comments. In addition:
- naming for `4_5b` and `4_5B` and `4-5b` should be consistent (including file names).
- some concern about the diif size for `halo2_gadgets/src/utilities/lookup_range_check.rs`

File: halo2_gadgets/src/ecc/chip/mul_fixed/short.rs [hunk 7]
```diff
@@ -792,156 +818,147 @@ pub mod tests {
...
+        ) -> Result<(), Error> {
+            let chip = EccChip::<TestFixedBases, Lookup>::construct(config.clone());
 
-                //let short_config = config.mul_fixed_short.clone();
-                let base = Point::new(chip, layouter.namespace(|| "load base"), self.base)?;
+            let column = config.advices[0];
 
-                let sign =
-                    self.load_private(layouter.namespace(|| "load sign"), column, self.sign)?;
+            //let short_config = config.mul_fixed_short.clone();
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply halo2_gadgets/src/ecc/chip/mul_fixed/short.rs:893]: Done

File: halo2_gadgets/src/sinsemilla.rs [hunk 8]
```diff
@@ -602,246 +585,346 @@ pub(crate) mod tests {
...
+
+            result.constrain_equal(
+                layouter.namespace(|| "result == expected result"),
+                &expected_result,
+            )
+        }
+    }
+
+    impl<Lookup: PallasLookupRangeCheck> Circuit<pallas::Base> for MyCircuit<Lookup> {
+        #[allow(clippy::type_complexity)]
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply halo2_gadgets/src/sinsemilla.rs:811]: Removed

File: halo2_gadgets/src/sinsemilla/chip.rs [hunk 9]
```diff
@@ -58,15 +59,18 @@ where
     /// generators of the Sinsemilla hash.
     pub(super) generator_table: GeneratorTableConfig,
     /// An advice column configured to perform lookup range checks.
-    lookup_config: LookupRangeCheckConfig<pallas::Base, { sinsemilla::K }>,
+    lookup_config: Lookup,
+
+    enable_hash_from_private_point: bool,
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply halo2_gadgets/src/sinsemilla/chip.rs:64]: Done

File: halo2_gadgets/src/sinsemilla/chip.rs [hunk 10]
```diff
@@ -153,15 +162,16 @@ where
         advices: [Column<Advice>; 5],
         witness_pieces: Column<Advice>,
         fixed_y_q: Column<Fixed>,
-        lookup: (TableColumn, TableColumn, TableColumn, TableColumn),
-        range_check: LookupRangeCheckConfig<pallas::Base, { sinsemilla::K }>,
+        lookup: (TableColumn, TableColumn, TableColumn),
+        range_check: Lookup,
+        enable_hash_from_private_point: bool,
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply halo2_gadgets/src/sinsemilla/chip.rs:167]: Done

File: halo2_gadgets/src/sinsemilla/merkle.rs (hunk 4 above)
Comment (ConstanceBeguier) [reply halo2_gadgets/src/sinsemilla/merkle.rs:622]: Done

File: halo2_gadgets/src/utilities/cond_swap.rs (hunk 5 above)
Comment (ConstanceBeguier) [reply halo2_gadgets/src/utilities/cond_swap.rs:443]: Done

File: halo2_gadgets/src/utilities/lookup_range_check.rs (hunk 6 above)
Comment (ConstanceBeguier) [reply halo2_gadgets/src/utilities/lookup_range_check.rs:717]: The diff between this PR and [PR to add a LookupRangeCheck trait](https://github.com/zcash/halo2/pull/823) is quite small

File: halo2_gadgets/src/ecc.rs [hunk 11]
```diff
@@ -967,7 +988,38 @@ pub(crate) mod tests {
...
+        let k = 13;
+        let circuit = MyCircuit::<PallasLookupRangeCheck45BConfig>::new(true);
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply halo2_gadgets/src/ecc.rs:1000]: Done

File: halo2_gadgets/src/ecc.rs (hunk 2 above)
Comment (ConstanceBeguier) [reply halo2_gadgets/src/ecc.rs:998]: Done

**Review by PaulLaux — APPROVED** [review]
Very nice

**Comment by PaulLaux** [comment]
can you clarify this: https://github.com/QED-it/halo2/pull/31#discussion_r1699615527

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ConstanceBeguier) [reply .github/workflows/ci.yml]: I removed these unnecessary changes

File: halo2_gadgets/src/sinsemilla/chip.rs [hunk 12]
```diff
@@ -330,6 +345,10 @@ where
         Q: &Self::NonIdentityPoint,
         message: Self::Message,
     ) -> Result<(Self::NonIdentityPoint, Vec<Self::RunningSum>), Error> {
+        if !self.config().enable_hash_from_private_point {
+            return Err(Error::HashFromPrivatePoint);
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply halo2_gadgets/src/sinsemilla/chip.rs]: Done

**Pull request merged**